### PR TITLE
Update inapp reviews flow

### LINKED ISSUE #20820: Enhance in-app reviews logic
Currently, we show an in-app review prompt after the second post has been published, and only once. However, we’re overlooking users who manage sites or only read without publishing. 

#### Required changes
* If “LATER” is tapped on the old dialog, do not show the new prompt for 7 days.
* If “NO THANKS” is tapped on the old dialog, do not show the new prompt at all.
* If “RATE NOW” is tapped on the old dialog, reset the counters of the new prompt.
* When the old dialog is shown, reset the counters of the new prompt.
* Keep the old method but focus only on the happiest moments for the new prompt, since we can’t ask “Enjoying the app?” with the new prompt. So, update the new prompt logic with:
  * (Already existing): The second post is published OR
  * “like”, “comment like” and “follow” notifications are tapped 10 times in total.
* Reset the counters and wait for 7 days after the in-app review prompt is shown. (If the user has already rated, the prompt will not show up but only Google can control this behavior.)

Link for internal discussion: pc8eDl-1nq-p2#comment-1088

### PR DESCRIPTION
**Pull request by irfano**
Fixes #20820 
This updates the logic of the in-app review prompt. 

#### Changes
* If “LATER” is tapped on the old dialog, do not show the new prompt for 7 days.
* If “NO THANKS” is tapped on the old dialog, do not show the new prompt at all.
* If “RATE NOW” is tapped on the old dialog, reset the counters of the new prompt.
* When the old dialog is shown, reset the counters of the new prompt.
* Keep the old method but focus only on the happiest moments for the new prompt, since we can’t ask “Enjoying the app?” with the new prompt. So, update the new prompt logic with:
  * (Already existing): The second post is published OR
  * “like”, “comment like” and “follow” notifications are tapped 10 times in total.
* Reset the counters and wait for 7 days after the in-app review prompt is shown. (If the user has already rated, the prompt will not show up but only Google can control this behavior.)

Screenshot 2024-05-28 at 11 24 39

> [!TIP]
> **For the reviewer:** I removed the `ReviewViewModel` and combined the logic of the in-app reviews prompt with the old rating dialog. I renamed `AppRatingDialog` to `AppReviewsManager` because using a manager class seems better than dealing with a view model. `AppReviewsManager` needs unit tests and some refactoring, such as using `AppPrefsWrapper` instead of `SharedPreference` in itself. However, I left these areas out of the PR's scope since the old rate dialog may be removed in the future.

-----

## To Test:

It's not possible to test this from the PR. It should be released to the store to be able to see the prompt. Launching the prompt was already tested in https://github.com/wordpress-mobile/WordPress-Android/pull/19741. Testing it again is not necessary. You can test the flow by checking debug logs. Logs can be viewed from Logcat or "Me → Help → Logs".

### Published post goal
1. If logged in, log out and log in again.
2. Publish a post from "My Site" or "My Site → Posts". 
3. Verify that you see `In-app reviews counter for published posts: 1` in the logs. 
4. Publish another post from "My Site" or "My Site → Posts".
5. Verify that you see `In-app reviews counter for published posts: 2` in logs. 
6. When it returns to the previous screen, verify that you see `Launching in-app reviews prompt` in logs. 

### Notification goal
1. If logged in, log out and log in again.
2. Send a notification to your app by following your selected site, commenting, and liking from another account on the web.
3. Tap the notification from the OS's system bar or "Notifications" tab within the app.
4. Verify that `In-app reviews counter for notification: 1` is in the logs.
5. Repeat step 3 and verify that the counter increases.
6. When navigating back after tapping the 10th notification, verify that you see `Launching in-app reviews prompt` in the logs. 

-----

## Regression Notes

1. Potential unintended areas of impact

    - The logic of the old dialog may be affected.

7. What I did to test those areas of impact (or what existing automated tests I relied on)

    - I manually reviewed the code change to ensure it isn't broken.

8. What automated tests I added (or what prevented me from doing so)

    - Updated current tests.

-----

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## Testing Checklist (strike-out the not-applying and unnecessary ones):

- [ ] ~WordPress.com sites and self-hosted Jetpack sites.~
- [ ] ~Portrait and landscape orientations.~
- [ ] ~Light and dark modes.~
- [ ] ~Fonts: Larger, smaller and bold text.~
- [ ] ~High contrast.~
- [ ] ~Talkback.~
- [ ] ~Languages with large words or with letters/accents not frequently used in English.~
- [ ] ~Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)~
- [ ] ~Large and small screen sizes. (Tablet and smaller phones)~
- [ ] ~Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)~

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java [hunk 1]
```diff
@@ -71,6 +65,8 @@
 import org.wordpress.android.fluxc.store.SiteStore.OnSiteChanged;
 import org.wordpress.android.fluxc.store.SiteStore.OnSiteEditorsChanged;
 import org.wordpress.android.fluxc.store.SiteStore.OnSiteRemoved;
+import org.wordpress.android.inappupdate.IInAppUpdateManager;
```
Comment (aditi-bhatia) [inline WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java:68]: Small nit: I know this is not a change you made, but is there a typo in the naming of `IInAppUpdateManager`?

File: WordPress/src/main/java/org/wordpress/android/widgets/AppReviewManager.kt [hunk 2]
```diff
@@ -94,6 +127,42 @@ object AppRatingDialog {
...
+        if (shouldShowInAppReviewsPrompt()) return
+        val shouldTrack = note.isUnread && (note.isLikeType || note.isCommentType || note.isFollowType)
+        if (shouldTrack && AppPrefs.getInAppReviewsNotificationCount() < TARGET_COUNT_NOTIFICATIONS) {
+            AppPrefs.incrementInAppReviewsNotificationCount()
+            AppLog.d(T.UTILS, "In-app reviews counter for notification: ${AppPrefs.getInAppReviewsNotificationCount()}")
+        }
+    }
+
+    /**
+     * Check whether the in-app reviews prompt should be shown or not.it after its last sh
```
Comment (aditi-bhatia) [inline WordPress/src/main/java/org/wordpress/android/widgets/AppReviewManager.kt]: Small nit: typo at the end of the sentence here

**Review by aditi-bhatia — APPROVED** [review]
Nice work!  

>  I renamed AppRatingDialog to AppReviewsManager because using a manager class seems better than dealing with a view model. 

I agree that the code feels cleaner this way ^. 

Code changes lgtm and testing works as expected 👍 I left 2 small nit-pick comments if you'd like to address those, otherwise good to merge

File: WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java (hunk 1 above)
Comment (irfano) [reply WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java:68]: I am not sure. @pantstamp, is there a typo in `IInAppUpdateManager`, or is it Interface's 'I' that we are used to seeing in C# code style?

File: WordPress/src/main/java/org/wordpress/android/widgets/AppReviewManager.kt (hunk 2 above)
Comment (irfano) [reply WordPress/src/main/java/org/wordpress/android/widgets/AppReviewManager.kt]: Good catch! Fixed: https://github.com/wordpress-mobile/WordPress-Android/pull/20891/commits/e4689656c17e6da20cd20b908155f4df1250e524

**Pull request merged**